### PR TITLE
FSE: Render dynamic blocks on page areas replaced by template parts

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

In order to render a dynamic block, WordPress uses [`do_blocks`](https://developer.wordpress.org/reference/functions/do_blocks/) to parse and replace it with the actual HTML code.

The post content is [automatically filtered by that function](https://github.com/WordPress/WordPress/blob/225de4e6eb6ef4ae0de4a6945aab6556433acad9/wp-includes/default-filters.php#L172), so any other block included in a different area like the site header or footer is not parsed by default.

Since the FSE Template Parts can be inserted out of the post content, we need to ensure that the dynamic blocks are parsed when we replace the page areas with a template part.

To do so, on this PR we're including a call to `do_blocks` before returning the replaced HTML.

I included other small changes that should improve the replacement logic:
- In order to keep intact the HTML loaded by `DOMDocument::loadHTML`, we use now the [`LIBXML_HTML_NOIMPLIED` option](https://www.php.net/manual/en/libxml.constants.php#constant.libxml-html-noimplied) which turns off an automatic clean process that was removing blocks comments markup (i.e. `<!-- a8c:site-logo /-->`) or adding extra `html`/`body` elements wrapping the template parts content.
- The above required to wrap any HTML loaded by `DOMDocument::loadHTML` in an element, since LibXML requires a root node and a template part could contain more than one node.
- Instead of replacing the `.site-header`/`.site-footer` elements with the template parts content, we keep maintaining the container, remove all the child nodes and the template part is appended as a new child:

| HTML before replacement | HTML after replacement (currently in master) | HTML after replacement (on this PR) |
|---|---|---|
| `<header class="site-header"><div class="site-branding-container">...</div></header>` | `<h2>Test Content Header</h2>` | `<header class="site-header"><h2>Test Content Header</h2></header>` |

#### Testing instructions

* Set up a local dev environment by following instructions available at PCYsg-ly5-p2.
* Create a new Template Part and insert a dynamic block (i.e. Site Logo) into it. Save it with "Header" as title.
* Create other new Template Part and insert a dynamic block (i.e. Recent Post) and a static block (i.e. Paragraph) into it. Save it with "Footer" as title.
* Create a new Template CPT with "Default Template" as title and add the following blocks into it:
  * Template Part (select "Header" from previous step).
  * Content Slot.
  * Template Part (select "Footer" from previous step).
* Create a test page and assign the previously created "Default Template" to it.

Screenshot 2019-06-07 at 03 19 23

* Publish the page and verify on the published view that the header and footer have been replaced with previously created template parts and the expected blocks are rendered.

Fixes https://github.com/Automattic/wp-calypso/issues/34166

### CODE REVIEW

**Comment by mmtr** [comment]
This is currently replacing the header as expected on my tests, but the footer is always empty on the published view.

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/utils/replace-template-parts.php [hunk 1]
```diff
@@ -58,11 +65,27 @@ function a8c_fse_replace_template_parts( $html ) {
...
 		$fse_node = $doc->importNode( $temp_doc->documentElement, true );
-		$node_to_replace->parentNode->replaceChild( $fse_node, $node_to_replace );
+
+		// Remove all the children of the element to be replaced in order to keep using the same container for the FSE content.
+		while ( $node_to_replace->hasChildNodes() ) {
+			$node_to_replace->removeChild( $node_to_replace->firstChild );
+		}
+
+		// Rather than appending the FSE content node, we add its child nodes so the extra wrapping div added above is not rendered.
+		for ( $i = 0; $i < $fse_node->childNodes->length; $i++ ) {
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/utils/replace-template-parts.php]: Seems this is the problematic code causing the weird behavior with the footer template parts I noted on https://github.com/Automattic/wp-calypso/pull/34171#issuecomment-504298538

Reverting this back to `$node_to_replace->parentNode->replaceChild( $fse_node, $node_to_replace );` solves the problem, but we have an extra `div` wrapping the template part content and we lose the `.site-header`/`.site-footer` elements.

Maybe that's not a big deal, so anyone is welcome to change it.

**Comment by kwight** [comment]
> In order to render a dynamic block, WordPress uses do_blocks to parse and replace it with the actual HTML code.

Ah, that was the magic plot point I was missing; I knew there must be some function with that duty.

**Comment by gwwar** [comment]
Hmm, I'm getting a blank screen on the publish views

**Comment by kwight** [comment]
@gwwar Same here, with fatals related to output buffering: `PHP Fatal error:  ob_start(): Cannot use output buffering in output buffering display handlers`. Trying to sort it out.

**Comment by gwwar** [comment]
Is this worth fixing, or should we start on our longer term approach? cc @rodrigoi @mattwiebe

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/utils/replace-template-parts.php (hunk 1 above)
Comment (kwight) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/utils/replace-template-parts.php]: Fixed this for now in 2503f698c1.

**Comment by gwwar** [comment]
When adding a site description block:
```
PHP Fatal error:  ob_start(): Cannot use output buffering in output buffering display handlers in /var/www/html/wp-content/plugins/full-site-editing-plugin/full-site-editing/blocks/site-description/index.php on line 14
```

A similar error also fires for site title

**Comment by kwight** [comment]
Ah, we're going to have this parsing problem for any dynamic blocks. The parsing uses output buffers, while our dynamic blocks also rely on them, so the buffer is trying to buffer another buffer (which is apparently not cool). Since we can't change the blocks themselves, moving on to an approach other than the parser would probably be a good idea.

**Comment by mmtr** [comment]
Closing this in favor of https://github.com/Automattic/wp-calypso/pull/34196

**Pull request closed without merging**